test/statement_types.py

```diff
@@ -0,0 +1,451 @@
+import numbers
+import os
+
+from sig_utils import overload_scalar, get_cpp_type, arg_types
+
+class CppStatement:
+    """Base class for all statements. Implements some default checks. Should not be used directly"""
+    def __init__(self):
+        raise Exception("CppStatement should never be instantiated directly")
+
+    def is_reverse_mode(self):
+        """By default, a statement is not reverse mode"""
+        return False
+    
+    def is_matrix_like(self):
```

Comment:
Could this be `is_eigen` or maybe `is_expression_compatible`?

Reply:
`is_eigen_compatible` seems fine and matches discussion [above](https://github.com/stan-dev/math/pull/2419/files#r595937988), so I can do that.